From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Paul Sauve <paul@technove.co>
Date: Sat, 31 Oct 2020 19:03:25 -0500
Subject: [PATCH] Queue lighting update only once

Airplane Lite
Copyright (C) 2020 Technove LLC

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.

diff --git a/src/main/java/net/minecraft/server/ChunkProviderServer.java b/src/main/java/net/minecraft/server/ChunkProviderServer.java
index 12d9b73ccc2f4406957932397746cac7902d650e..fc9919dd01cad02486af293f9b0c692046c7eea1 100644
--- a/src/main/java/net/minecraft/server/ChunkProviderServer.java
+++ b/src/main/java/net/minecraft/server/ChunkProviderServer.java
@@ -1005,6 +1005,7 @@ public class ChunkProviderServer extends IChunkProvider {
             // Paper - moved up
             // Tuinity start - optimise chunk tick iteration
             com.tuinity.tuinity.util.maplist.IteratorSafeOrderedReferenceSet.Iterator<Chunk> iterator = this.entityTickingChunks.iterator();
+            boolean updateLighting = false; // AirplaneL
             try {
             while (iterator.hasNext()) {
                 Chunk chunk = iterator.next();
@@ -1029,7 +1030,7 @@ public class ChunkProviderServer extends IChunkProvider {
                             }
 
                             this.world.timings.chunkTicks.startTiming(); // Spigot // Paper
-                            this.world.a(chunk, k);
+                            if (this.world.abool(chunk, k)) updateLighting = true; // AirplaneL
                             this.world.timings.chunkTicks.stopTiming(); // Spigot // Paper
                             MinecraftServer.getServer().executeMidTickTasks(); // Tuinity - exec chunk tasks during world tick
                         }
@@ -1039,6 +1040,7 @@ public class ChunkProviderServer extends IChunkProvider {
             } finally {
                 iterator.finishedIterating();
             }
+            if (updateLighting) this.getLightEngine().queueUpdate(); // AirplaneL
             // Tuinity end - optimise chunk tick iteration
             this.world.getMethodProfiler().enter("customSpawners");
             if (flag1) {
diff --git a/src/main/java/net/minecraft/server/WorldServer.java b/src/main/java/net/minecraft/server/WorldServer.java
index 72452abfe1fa8eafac386e0f2e5d87296772675a..bf1502c8e9277b533933b43a453b98c3b09b7849 100644
--- a/src/main/java/net/minecraft/server/WorldServer.java
+++ b/src/main/java/net/minecraft/server/WorldServer.java
@@ -978,7 +978,10 @@ public class WorldServer extends World implements GeneratorAccessSeed {
     private final com.destroystokyo.paper.util.math.ThreadUnsafeRandom randomTickRandom = new com.destroystokyo.paper.util.math.ThreadUnsafeRandom();
     // Paper end
 
-    public void a(Chunk chunk, int i) { final int randomTickSpeed = i; // Paper
+    // AirplaneL start - create version of chunk tick that returns a bool for updating lighting
+    public void a(Chunk chunk, int i) { this.abool(chunk, i); }
+    public boolean abool(Chunk chunk, int i) { final int randomTickSpeed = i; // Paper
+        // AirplaneL end
         ChunkCoordIntPair chunkcoordintpair = chunk.getPos();
         boolean flag = this.isRaining();
         int j = chunkcoordintpair.d();
@@ -1081,9 +1084,13 @@ public class WorldServer extends World implements GeneratorAccessSeed {
             }
             gameprofilerfiller.exit();
             timings.chunkTicksBlocks.stopTiming(); // Paper
-            getChunkProvider().getLightEngine().queueUpdate(); // Paper
+            // AirplaneL start
+            //getChunkProvider().getLightEngine().queueUpdate(); // Paper
+            return true;
+            // AirplaneL end
             // Paper end
         }
+        return false; // AirplaneL
     }
 
     protected BlockPosition a(BlockPosition blockposition) {
